contracts/utility/CarbonBatcher.sol

```diff
@@ -0,0 +1,224 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.19;
+
+import { ReentrancyGuard } from "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import { IERC721Receiver } from "@openzeppelin/contracts/token/ERC721/IERC721Receiver.sol";
+import { Address } from "@openzeppelin/contracts/utils/Address.sol";
+
+import { ICarbonController } from "../carbon/interfaces/ICarbonController.sol";
+import { IVoucher } from "../voucher/interfaces/IVoucher.sol";
+
+import { Upgradeable } from "./Upgradeable.sol";
+import { Order } from "../carbon/Strategies.sol";
+import { Utils } from "../utility/Utils.sol";
+import { Token } from "../token/Token.sol";
+
+struct StrategyData {
+    Token[2] tokens;
+    Order[2] orders;
+}
+
+/**
+ * @dev Contract to batch create carbon controller strategies
+ */
+contract CarbonBatcher is Upgradeable, Utils, ReentrancyGuard, IERC721Receiver {
+    using Address for address payable;
+
+    error InsufficientNativeTokenSent();
+
+    ICarbonController private immutable carbonController;
+    IVoucher private immutable voucher;
+
+    /**
+     * @dev triggered when tokens have been withdrawn from the carbon batcher
+     */
+    event FundsWithdrawn(Token indexed token, address indexed caller, address indexed target, uint256 amount);
+
+    constructor(
+        ICarbonController _carbonController,
+        IVoucher _voucher
+    ) validAddress(address(_carbonController)) validAddress(address(_voucher)) {
+        carbonController = _carbonController;
+        voucher = _voucher;
+        _disableInitializers();
+    }
+
+    /**
+     * @dev fully initializes the contract and its parents
+     */
+    function initialize() public initializer {
+        __CarbonBatcher_init();
+    }
+
+    // solhint-disable func-name-mixedcase
+
+    /**
+     * @dev initializes the contract and its parents
+     */
+    function __CarbonBatcher_init() internal onlyInitializing {
+        __Upgradeable_init();
+    }
+
+    /**
+     * @inheritdoc Upgradeable
+     */
+    function version() public pure virtual override(Upgradeable) returns (uint16) {
+        return 1;
+    }
+
+    /**
+     * @notice creates several new strategies, returns the strategies id's
+     *
+     * requirements:
+     *
+     * - the caller must have approved the tokens with assigned liquidity in the orders
+     */
+    function batchCreate(
+        StrategyData[] calldata strategies
+    ) external payable greaterThanZero(strategies.length) nonReentrant returns (uint256[] memory) {
+        uint256[] memory strategyIds = new uint256[](strategies.length);
+        uint256 txValueLeft = msg.value;
+
+        // extract unique tokens and amounts
+        (Token[] memory uniqueTokens, uint256[] memory amounts) = _extractUniqueTokensAndAmounts(strategies);
+        // transfer funds from user for strategies
+        for (uint256 i = 0; i < uniqueTokens.length; i = uncheckedInc(i)) {
```

Comment:
Perhaps it's time to upgrade to Solidity >0.8.22? `uncheckedInc(i)` will be no longer needed, for example